Content.Server/Disease/DiseaseSystem.cs

```diff
@@ -0,0 +1,420 @@
+using Content.Server.Popups;
+using Content.Server.DoAfter;
+using Robust.Shared.Player;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+using Robust.Shared.Serialization.Manager;
+using Content.Shared.Inventory.Events;
+using Content.Server.Nutrition.EntitySystems;
+
+namespace Content.Server.Disease
+{
+
+    /// <summary>
+    /// Handles disease propagation & curing
+    /// </summary>
+    public sealed class DiseaseSystem : EntitySystem
+    {
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly ISerializationManager _serializationManager = default!;
+        [Dependency] private readonly IRobustRandom _random = default!;
+        [Dependency] private readonly DoAfterSystem _doAfterSystem = default!;
+        [Dependency] private readonly PopupSystem _popupSystem = default!;
+        [Dependency] private readonly EntityLookupSystem _lookup = default!;
+        [Dependency] private readonly SharedInteractionSystem _interactionSystem = default!;
+        [Dependency] private readonly InventorySystem _inventorySystem = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<DiseaseCarrierComponent, CureDiseaseAttemptEvent>(OnTryCureDisease);
+            SubscribeLocalEvent<DiseasedComponent, InteractHandEvent>(OnInteractDiseasedHand);
+            SubscribeLocalEvent<DiseasedComponent, InteractUsingEvent>(OnInteractDiseasedUsing);
+            SubscribeLocalEvent<DiseaseProtectionComponent, GotEquippedEvent>(OnEquipped);
+            SubscribeLocalEvent<DiseaseProtectionComponent, GotUnequippedEvent>(OnUnequipped);
+            SubscribeLocalEvent<DiseaseVaccineComponent, AfterInteractEvent>(OnAfterInteract);
+            SubscribeLocalEvent<DiseaseVaccineComponent, ExaminedEvent>(OnExamined);
+            /// Private events stuff
+            SubscribeLocalEvent<TargetVaxxSuccessfulEvent>(OnTargetVaxxSuccessful);
+            SubscribeLocalEvent<VaxxCancelledEvent>(OnVaxxCancelled);
+        }
+
+        private Queue<EntityUid> AddQueue = new();
+        private Queue<(DiseaseCarrierComponent, DiseasePrototype)> CureQueue = new();
+
+        /// <summary>
+        /// First, adds or removes diseased component from the queues and clears them.
+        /// Then, iterates over every diseased component to check for their effects
+        /// and cures
+        /// </summary>
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+            foreach (var entity in AddQueue)
+                EnsureComp<DiseasedComponent>(entity);
+            AddQueue.Clear();
+
+            foreach (var tuple in CureQueue)
+            {
+                if (tuple.Item1.Diseases.Count == 1) //This is reliable unlike testing Count == 0 right after removal for reasons I don't quite get
+                    RemComp<DiseasedComponent>(tuple.Item1.Owner);
```

Comment:
if you give the tuple args names this can be a little more obvious what it means